cpp/src/arrow/compute/kernels/vector_selection_take_internal.cc

```diff
@@ -715,12 +853,244 @@ class TakeMetaFunction : public MetaFunction {
+    } else if (values_chunked->num_chunks() == 1) {
+      single_chunk = values_chunked->chunk(0);
+    }
+
+    if (args[1].kind() == Datum::ARRAY) {
+      // CA->C
+      auto& indices = args[1].array();
+      if (single_chunk) {
+        // AA->A
+        DCHECK_EQ(single_chunk->length(), batch.length);
+        // If the ChunkedArray was cheaply converted to a single chunk,
+        // we can use the AA->A take kernel directly.
+        RETURN_NOT_OK(
+            CallAAAKernel(take_aaa_exec, ctx, single_chunk->data(), indices, out));
+        out->value = std::make_shared<ChunkedArray>(MakeArray(out->array()));
+        return Status::OK();
+      }
+      // Instead of concatenating the chunks, we call the CA->A take kernel
+      // which has a more efficient implementation for this case. At this point,
+      // that implementation doesn't have to care about empty or single-chunk
+      // ChunkedArrays.
+      RETURN_NOT_OK(take_caa_exec(ctx, batch, &result));
+      out->value = std::make_shared<ChunkedArray>(MakeArray(result.array()));
+      return Status::OK();
+    } else {
+      // CC->C
+      const auto& indices = args[1].chunked_array();
+      std::vector<std::shared_ptr<Array>> new_chunks;
+      for (int i = 0; i < indices->num_chunks(); i++) {
+        auto& indices_chunk = indices->chunk(i)->data();
+        result = PrepareOutput(batch, values_chunked->length());
+        if (single_chunk) {
+          // If the ChunkedArray was cheaply converted to a single chunk,
+          // we can use the AA->A take kernel directly.
+          RETURN_NOT_OK(CallAAAKernel(take_aaa_exec, ctx, single_chunk->data(),
+                                      indices_chunk, &result));
+        } else {
+          RETURN_NOT_OK(
+              CallCAAKernel(take_caa_exec, ctx, values_chunked, indices_chunk, &result));
+        }
+        new_chunks.push_back(MakeArray(result.array()));
+      }
+      DCHECK(out->is_array());
+      out->value =
+          std::make_shared<ChunkedArray>(std::move(new_chunks), values_chunked->type());
+      return Status::OK();
+    }
+  } else {
+    // VectorKernel::exec_chunked are only called when at least one of the inputs is
+    // chunked, so we should be able to assume that args[1] is a chunked array when
+    // everything is wired up correctly.
+    if (args[1].kind() == Datum::CHUNKED_ARRAY) {
+      // AC->C
+      return TakeACCChunkedExec(take_aaa_exec, ctx, batch, out);
+    } else {
+      DCHECK(false) << "Unexpected kind for array_take's exec_chunked kernel: values="
+                    << args[0].ToString() << ", indices=" << args[1].ToString();
+    }
+  }
+  return Status::NotImplemented(
+      "Unsupported kinds for 'array_take', try using 'take': "
+      "values=",
+      args[0].ToString(), ", indices=", args[1].ToString());
+}
+
+template <ArrayKernelExec kTakeAAAExec, VectorKernel::ChunkedExec kTakeCAAExec>
```

Comment:
Why is template used here since it delegate implement to SpecialTakeChunkedExec directly?

Reply:
For convenience of instantiation in `PopulateTakeKernels`.

`SpecialTakeChunkedExecFunctor::Exec` gets inlined but `SpecialTakeChunkedExec` does not. We avoid binary bloat because I don't really want `SpecialTakeChunkedExec` to be specialized on every combination of `kTakeAAAExec` and `kTakeCAAExec` — I immediately force them into becoming function pointers that I dispatch to at runtime (from `SpecialTakeChunkedExec`) instead of having them inlined.